File: Dynamic Programming/Coin Change 2/solution.ts
// Can use JSON.stringify array to be key but this is optimized version

function change(amount: number, coins: number[]): number {
    let map = new Map()
    function helper(amount: number, index: number): number {
        let key = `${amount}-${index}`
        if(map.has(key)) {
            return map.get(key)
        }
        if(amount === 0) {
            return 1
        }
        if(amount < 0 || index === coins.length) {
            return 0
        }
        map.set(key, helper(amount - coins[index], index) + helper(amount, index + 1))
        return map.get(key)
    }
    return helper(amount, 0)
};

File: Dynamic Programming/Coin Change 2/solution.go
package main

import "fmt"

func change(amount int, coins []int) int {
	cache := make(map[string]int)
	return helper(amount, coins, 0, cache)
}

func helper(amount int, coins []int, index int, cache map[string]int) int {
	key := fmt.Sprintf("%v-%v", amount, index)
	if _, ok := cache[key]; ok {
		return cache[key]
	}
	if amount == 0 {
		return 1
	}
	if amount < 0 || index == len(coins) {
		return 0
	}
	cache[key] = helper(amount-coins[index], coins, index, cache) + helper(amount, coins, index+1, cache)
	return cache[key]
}
